### ISSUE TITLE
System.InvalidOperationException: Number of initialized language workers exceeded:11 exceeded maxProcessCount: 8

### ISSUE THREAD

**Issue by blacelle** [Azure/azure-functions-host#9335]
#### Investigative information

This issue seems very related with: https://github.com/Azure/azure-functions-host/issues/4559

Please provide the following:

- Timestamp: timestamp [UTC] 2023-06-15T19:32:04.5281845Z
- Function App version:
- Function App name: prd-retriever-engine
- Function name(s) (as appropriate):
- Invocation ID: 12f712bb-7e55-4655-820a-da966a6bce5b
- Region: france-central

#### Repro steps

We do not know how to reproduce it. It happens more regularly since a few days, after upgrade Functions

#### Expected behavior

No exception, preventing Functions invokations, and corrupting the worker until restarted.

#### Actual behavior

We observe stack like:
```
retriever-68ce089b-7a70-4032-8732-c8477a28837a: User message handler complete: Message: SequenceNumber: 866757, Exception: Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Exception while executing function: Functions.function_scraper
---> System.InvalidOperationException: Number of initialized language workers exceeded:11 exceeded maxProcessCount: 8
at Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcFunctionInvocationDispatcher.GetInitializedWorkerChannelsAsync(String language) in /src/azure-functions-host/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:line 440
at Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcFunctionInvocationDispatcher.InvokeAsync(ScriptInvocationContext invocationContext) in /src/azure-functions-host/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:line 399
at Microsoft.Azure.WebJobs.Script.Description.WorkerFunctionInvoker.InvokeCore(Object[] parameters, FunctionInvocationContext context) in /src/azure-functions-host/src/WebJobs.Script/Description/Workers/WorkerFunctionInvoker.cs:line 97
at Microsoft.Azure.WebJobs.Script.Description.FunctionInvokerBase.Invoke(Object[] parameters) in /src/azure-functions-host/src/WebJobs.Script/Description/FunctionInvokerBase.cs:line 82
at Microsoft.Azure.WebJobs.Host.Executors.VoidTaskMethodInvoker`2.InvokeAsync(TReflected instance, Object[] arguments) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\VoidTaskMethodInvoker.cs:line 20
at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:line 52
at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeWithTimeoutAsync(IFunctionInvoker invoker, ParameterHelper parameterHelper, CancellationTokenSource timeoutTokenSource, CancellationTokenSource functionCancellationTokenSource, Boolean throwOnTimeout, TimeSpan timerInterval, IFunctionInstance instance) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 581
at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstanceEx instance, ParameterHelper parameterHelper, ILogger logger, CancellationTokenSource functionCancellationTokenSource) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 527
at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance, FunctionStartedMessage message, FunctionInstanceLogEntry instanceLogEntry, ParameterHelper parameterHelper, ILogger logger, CancellationToken cancellationToken) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 306
--- End of inner exception stack trace ---
at Microsoft.Azure.WebJobs.ServiceBus.MessageProcessor.CompleteProcessingMessageAsync(ServiceBusMessageActions actions, ServiceBusReceivedMessage message, FunctionResult result, CancellationToken cancellationToken)
at Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusListener.ProcessMessageAsync(ProcessMessageEventArgs args)
at Azure.Messaging.ServiceBus.ServiceBusProcessor.OnProcessMessageAsync(ProcessMessageEventArgs args)
at Azure.Messaging.ServiceBus.ReceiverManager.OnMessageHandler(EventArgs args)
at Azure.Messaging.ServiceBus.ReceiverManager.ProcessOneMessage(ServiceBusReceivedMessage triggerMessage, CancellationToken cancellationToken), LockToken: c9ffa449-5b9d-44c7-b592-18791337772a
```

#### Known workarounds

Force-restart the Function when the issue kicks-in.

#### Related information 

Provide any related information 

* worker language: Python

Related host.json:

```
{
  "version": "2.0",
  "extensionBundle": {
    "id": "Microsoft.Azure.Functions.ExtensionBundle",
    "version": "[4.0.0, 5.0.0)"
  },
  "extensions": {
    "serviceBus": {
      "clientRetryOptions": {
        "mode": "Fixed",
        "tryTimeout": "00:15:00",
        "delay": "00:00:00.80",
        "maxDelay": "00:00:05",
        "maxRetries": 3
      },
      "prefetchCount": 0,
      "maxAutoLockRenewalDuration": "00:15:00",
      "maxConcurrentCalls": 8,
      "maxMessageBatchSize": 1
    }
  },
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": 3,
    "delayInterval": "00:00:05"
  },
  "logging": {
    "fileLoggingMode": "debugOnly",
    "logLevel": {
      "default": "Debug"
    },
    "console": {
      "isEnabled": true,
      "DisableColors": true
    }
  },
  "functionTimeout": "00:10:00"
}
```

Related function.json:
```
{
  "scriptFile": "__init__.py",
  "bindings": [
    {
      "name": "msg",
      "authLevel": "anonymous",
      "type": "serviceBusTrigger",
      "direction": "in",
      "queueName": "retriever",
      "connection": "AzureWebJobsServiceBus"
    }
  ]
}
```

**Comment by guillaumepichenot** [comment]
In case this would help your teams investigate, we had multiple occurrences of this (same app as @blacelle).

One more remarkable than others was : 
- Function App name: prd-retriever-engine
- Function name : function_scrapper
- Host Instance ID : 9e113eb8-8329-4fa9-b1b0-ffe8bd341bf6

What is remarkable : 
1. AppInsights' first trace for this HostInstanceID is at 2023-06-27T19:06:14Z
2. The first invocation on this instance is 3ad66dcb-e936-498f-b729-947d1b0c3e03, logged in traces at 2023-06-27T19:09:39Z (~3 minutes after HostInstanceID creation)
3. This invocation fails immediately with exception "Number of initialized language workers exceeded:10 exceeded maxProcessCount: 8" (exception itemId = `2d5cc267-151e-11ee-bac3-b578145ab334`)

So this instance was failing before it even reached applicative code, in a way.

**Comment by liliankasem** [comment]
Sorry it has taken me so long to get to this, is this issue still happening with the latest host version? If so, do you mind sharing a new timestamp for me to investigate? Thanks!

**Comment by guillaumepichenot** [comment]
No recent occurence of this issue as of late.

We worked with Azure support on this, and discovered something that seemed to be the trigger : as part of the standard python behavior, some files were written by the python runtime in `__pycache__` folders.

This would trigger a "restart" of the function runtime on the host (a log in App insight that I can not exactly quote by memory says that when files are written under a certain directory, it triggers a restart).

Quoting our discussion with Azure support : 
 > The above behavior (/__pycache__) would cause different unexpected issues as mentioned in [Function host stopping in the middle of function execution · Issue #7182 · Azure/azure-functions-host (github.com)](https://github.com/Azure/azure-functions-host/issues/7182) and I assume the other issues you encountered could be related to this as well. Would you please follow [__pycache__ creation causing restarts · Issue #366 · Azure/azure-functions-python-worker (github.com)](https://github.com/Azure/azure-functions-python-worker/issues/366) to avoid the “__pycache__” folder being created unexpectedly?

Following this discussion, we added a non-empty `PYTHONDONTWRITEBYTECODE` environment variable to our host fuction, which in turn disabled `__pycache__` folder creation, and the restarts disappeared.
Subsequently, no occurence of the `Number of initialized language workers exceeded` error appeared.


From an availability point of view, I consider the issue resolved. Our app function's runtime is not getting into the error state, and we do not need to reboot it anymore it's fine.


Nonetheless, this looks very surprising. The creation of pycache folders is a standard python feature, and it is enabled by default on the azure function runtime. If it can somehow lead to a dysfunctional host, there probably is a misbehaving component in the stack.

From an app developer's perspective, all this is non customisable, azure internal code and infrastructure.

The very fact that a standard applications behavior (python process creating pycache folder) should trigger a function runtime restart (which costs a few seconds at the very least) looks very strange to me in a server less (and supposedly low-latency) stack.

So something is off. Exactly what escaped us - even though we worked around it.

**Comment by green-digger** [comment]
We were having the same issue.

The Python Function App would not be able to start new instances due to the  "Number of initialized language workers exceeded:..." exception.

We tested a number of hypotheses:

1. the function initialisation was too slow
2. the resource allocation in terms of instance size and number of workers was incorect.
3. too high concurrency

None of these resolved the issue.

We then tried the solution in [here](https://github.com/Azure/azure-functions-host/issues/9335#issuecomment-1750139305) with many thanks to @guillaumepichenot 

The turning off of __pycache__ removed the problem.

@liliankasem I don't think this issue should be closed. At a minimum I think the disabled pycache solution should be elevated to the main documentation to ensure others can find it.

**Comment by liliankasem** [comment]
We got a CRI for this and have been using it to investigate; we have some ideas on how to address this and will look into if a hotfix will be required.

Essentially what is happening is that we create channels up till the maxProcessCount, even if the JobHost has exited. The dispatcher is not checking any current state of channel to know if it should continue. So if we have a JobHost restart (due to something like a file change), and we never dispose this loop, we'll have two JobHosts both going 0 -> maxProcessCount. They both add to the same "pool" of workers that are at the WebHost-level now that Worker Indexing is enabled (as of 4.28, worker indexing is enabled by default).

With Worker Indexing disabled the same thing happens, but they are scoped and isolated. So JobHost A will create 3 and so will JobHost B -- but when we go to check the active workers (which is where this exception is being thrown), we'll only see those from the current JobHost, none will be at the WebHost level and so we don't get that exception.

All to say, I believe the pycache relation here is a bit of a red herring - there is likely activity (i.e. folder/file creation or deletion) that is leading to a JobHost restart and eventually this issue.

**Comment by green-digger** [comment]
Thanks for reply @liliankasem. Really appreciate the explanation.

In general, should we prevent folder/file creation to prevent the host from restarting? Is there any documentation that I can read on this?

**Comment by liliankasem** [comment]
Manipulating files/folders should be fine outside of the function folder, we have watched directories where if we see changes we restart the host. I haven't been able to find much docs other than this:

https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-csharp?tabs=functionsv2%2Cfixed-delay%2Cazure-cli#watched-directories

And there's a code comment that explains this a bit more - we should make sure that this info is available in public docs, I will look into it next week.

https://github.com/Azure/azure-functions-host/blob/d6653fb45d7b92fed8914a7f0f334e3fc54376b4/src/WebJobs.Script.WebHost/FileMonitoringService.cs#L268-L274

**Comment by liliankasem** [comment]
Bug fix has been merged and should be available in v4.30 of the host (ETA mid Feb)
